dev/Notifications/BadgeNotifications/BadgeNotificationManager.cpp

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) Microsoft Corporation and Contributors.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "BadgeNotificationManager.h"
+#include "Microsoft.Windows.BadgeNotifications.BadgeNotificationManager.g.cpp"
+#include "externs.h"
+#include <frameworkudk/toastnotifications.h>
+#include <FrameworkUdk/toastnotificationsrt.h>
+#include <Microsoft.RoApi.h>
+#include "BadgeNotification.h"
+#include "BadgeNotificationTelemetry.h"
+
+using namespace std::literals;
+
+namespace winrt
+{
+    using namespace winrt::Windows::UI;
+    using namespace winrt::Windows::Foundation;
+    using namespace winrt::Windows::Foundation::Collections;
+    using namespace winrt::Microsoft::Windows::AppNotifications;
+    using namespace Windows::ApplicationModel::Core;
+    using namespace winrt::Microsoft::Windows::BadgeNotifications;
+}
+
+namespace ToastABI
+{
+    using namespace ::ABI::Microsoft::Internal::ToastNotifications;
+    using namespace ::ABI::Windows::Foundation::Collections;
+}
+
+namespace BaseNotifications
+{
+    using namespace ::Microsoft::Windows::BaseNotifications;
+}
+
+namespace winrt::Microsoft::Windows::BadgeNotifications::implementation
+{
+    BadgeNotificationManager::BadgeNotificationManager() : BaseNotifications::BaseNotificationManager() {}
+
+    winrt::Microsoft::Windows::BadgeNotifications::BadgeNotificationManager BadgeNotificationManager::Current()
+    {
+        auto appProperties{ winrt::CoreApplication::Properties() };
+
+        static wil::srwlock lock;
+
+        auto criticalSection{ lock.lock_exclusive() };
+        auto storedBadgeNotificationManager{ appProperties.TryLookup(STORED_BADGENOTIFICATION_MANAGER_KEY) };
+        if (storedBadgeNotificationManager)
+        {
+            return storedBadgeNotificationManager.as<winrt::Microsoft::Windows::BadgeNotifications::BadgeNotificationManager>();
+        }
+        else
+        {
+            // Need to clear the RoActivateInstance caching for the PushNotificationLongRunningProcess proxyStub to be found.
+            ::Microsoft::RoApi::ClearRoActivateInstanceCache();
+
+            // Store the BadgeNotificationManager in the COM static store
+            auto badgeNotificationManager{ winrt::make<BadgeNotificationManager>() };
+            appProperties.Insert(STORED_BADGENOTIFICATION_MANAGER_KEY, badgeNotificationManager);
+            return badgeNotificationManager;
+        }
+    }
+
+    void BadgeNotificationManager::SetBadgeAsCount(uint32_t notificationCount)
+    {
+        THROW_HR_IF_MSG(E_ILLEGAL_METHOD_CALL, !AppModel::Identity::IsPackagedProcess(), "Not applicable for unpackaged applications");
+
+        SetBadge(std::to_wstring(notificationCount), nullptr);
+        return;
+    }
+
+    void BadgeNotificationManager::SetBadgeAsCount(uint32_t notificationCount, winrt::DateTime expiration)
```

Comment:
validate these arguments

Reply:
We don't need to validate these arguments because shell takes care of this. For e.g if the count is 0 no badge gets displayed if it is >100 only 99+ gets displayed. This behavior is as same as in current WindowsSdk implementation